### PR TITLE
Validate maximum length of strings in models (before reaching DB)

### PR DESCRIPTION
**Pull request by Martouta**
Fixes [THREESCALE-2367](https://issues.jboss.org/browse/THREESCALE-2367)
It validates for all models that the length of the strings is not higher than what the field can take, so it doesn't reach the DB and it doesn't raise an error for requests through the API but responds with a nice error message instead.

### CODE REVIEW

File: app/models/proxy_config.rb [hunk 1]
```diff
@@ -25,6 +25,7 @@ class InvalidEnvironmentError < StandardError; end
   validates :environment, inclusion: { in: ENVIRONMENTS }
   validate :service_token_exists
   validates :content, length: { maximum: MAX_CONTENT_LENGTH }
+  # validates :environment, length: { maximum: 255 }
```
Comment (Martouta) [inline app/models/proxy_config.rb]: WIP 😄

File: app/models/proxy_rule.rb [hunk 2]
```diff
@@ -89,8 +89,13 @@ def querystring_parameters
     Hash[query.split('&').map { |kv| kv.split('=', 2) }]
   end
 
+  # TODO: temporarily, until I figure out why we keep this column in 'columns'
+  def metric_system_name=(metric_system_name)
```
Comment (Martouta) [inline app/models/proxy_rule.rb]: All this metric_system_name thing is just a WIP 😉

File: Gemfile.base [hunk 3]
```diff
@@ -124,6 +124,7 @@ gem 'diff-lcs', '~> 1.2'
 gem 'acts-as-taggable-on', '~> 4.0'
 gem 'whenever', '~> 0.9.7', require: false
 gem 'will_paginate', '~> 3.1.0'
+gem 'enum_attributes_validation', git: 'https://github.com/CristiRazvi/enum_attributes_validation'
```
Comment (Martouta) [inline Gemfile.base]: Temporarily I think 😄

File: app/models/user_session.rb [hunk 4]
```diff
@@ -7,6 +7,7 @@ class UserSession < ApplicationRecord
   has_one :account, through: :user
 
   validates :user_id, :key, presence: true
+  validates :ip, length: {maximum: 255}
```
Comment (josemigallas) [inline app/models/user_session.rb]: IP that long?

File: test/unit/models_test.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        column_sql_type = column.sql_type
+        next if column_sql_type.match(/\Acharacter varying\Z/)
+        length = column_sql_type.match(/\(([\d]+)\)/)[1].to_i
+
+        object = model.new({column_name => ('a' * (length + 1))}, without_protection: true)
+        object.valid?
+
+        column_errors = object.errors[column_name].to_sentence
+        next if column_errors.match(/is not included in the list/)
+        assert_match /too long/, column_errors, "#{model} is not validating the max length of #{length} for string '#{column_name}'"
```
Comment (josemigallas) [inline test/unit/models_test.rb]: You should also assert that it actually passes if <= length (And perhaps also = length)

File: app/models/user_session.rb (hunk 4 above)
Comment (josemigallas) [inline app/models/user_session.rb]: IP that long? Isn't there a type for ips?

Comment (Martouta) [reply app/models/user_session.rb]: In the schema it is that long :
https://github.com/3scale/porta/blob/4b1d9bf7d223a27e74afc81fcc3a1612d1447f6c/db/schema.rb#L1392

File: test/unit/models_test.rb (hunk 5 above)
Comment (Martouta) [reply test/unit/models_test.rb]: @josemigallas Not necessarily imo 🤔 Maybe sometimes we want a validation of the length that is less than what a string can take, that makes sense from the business logic. So all my goal here is not to make it bigger than what the DB can take, but if we want to enforce it to be less in any case, that's fine for me, i don't see why that should be a problem 😄

File: app/models/user_session.rb (hunk 4 above)
Comment (Martouta) [reply app/models/user_session.rb]: And we have another ip that we do the same:
https://github.com/3scale/porta/blob/4b1d9bf7d223a27e74afc81fcc3a1612d1447f6c/db/schema.rb#L1435

**Pull request merged**